packages/adblocker/src/engine/bucket/preprocessor.ts

```diff
@@ -0,0 +1,149 @@
+import { StaticDataView } from '../../data-view';
+import IFilter from '../../filters/interface';
+import Preprocessor, { Env } from '../../preprocessor';
+
+export default class PreprocessorBucket {
+  public static deserialize(view: StaticDataView): PreprocessorBucket {
+    const preprocessors: Preprocessor[] = [];
+
+    for (let i = 0, l = view.getUint32(); i < l; i++) {
+      preprocessors.push(Preprocessor.deserialize(view));
+    }
+
+    return new this({
+      preprocessors,
+    });
+  }
+
+  private readonly preprocessors: Preprocessor[];
+  private readonly excluded: Set<number>;
+
+  constructor({
+    env = new Env(),
+    preprocessors = [],
+  }: {
+    env?: Env;
+    preprocessors?: Preprocessor[];
+  }) {
+    this.excluded = new Set();
+    this.preprocessors = preprocessors;
+
+    if (preprocessors.length) {
+      this.update({ added: preprocessors, env });
+    }
+  }
+
+  private build(env: Env) {
+    // Update excluded filter ids based on bindings
+    this.excluded.clear();
+
+    const allowed = new Set<number>();
+
+    for (let i = 0; i < this.preprocessors.length; ) {
+      if (this.preprocessors[i].filterIDs.size === 0) {
+        this.preprocessors.splice(i, 1);
+
+        continue;
+      }
+
+      if (this.preprocessors[i].evaluate(env)) {
+        for (const filterID of this.preprocessors[i].filterIDs) {
+          allowed.add(filterID);
+        }
+      } else {
+        for (const filterID of this.preprocessors[i].filterIDs) {
+          if (!allowed.has(filterID)) {
+            this.excluded.add(filterID);
+          }
+        }
+      }
```

Comment:
It's unclear what we are solving here, are we giving priority to filters allowed as part of a block which would be seen before the same filter in an excluded block? The order is not guaranteed so that seems arbitrary. If there is some priority to apply we should do it separately: first collect all allowed filters and all excluded filtered and then resolve priority on the two final lists, instead of proceeding iteratively.

Reply:
I'd like to give priority to 'allowed' filters since we need to allow the filter if there is any of preprocessors evaluated to 'true'.